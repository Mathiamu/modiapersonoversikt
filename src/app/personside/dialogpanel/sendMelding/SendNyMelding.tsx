import * as React from 'react';
import { FormEvent, useEffect, useState } from 'react';
import { RadioPanelGruppe, Textarea } from 'nav-frontend-skjema';
import { Meldingstype } from '../../../../models/meldinger/meldinger';
import { Kodeverk } from '../../../../models/kodeverk';
import { UnmountClosed } from 'react-collapse';
import KnappBase from 'nav-frontend-knapper';
import styled from 'styled-components';
import Temavelger from '../component/Temavelger';
import { SkjemaelementFeil } from 'nav-frontend-skjema/lib/skjemaelement-feilmelding';
import KnappMedBekreftPopup from '../../../../components/KnappMedBekreftPopup';
import { useDispatch, useSelector } from 'react-redux';
import { sendMeldingActionCreator } from '../../../../redux/restReducers/sendMelding';
import { AppState } from '../../../../redux/reducers';
import { isLoaded, isNotStarted } from '../../../../rest/utils/restResource';
import { JournalforingsSak } from '../../infotabs/meldinger/traadvisning/verktoylinje/journalforing/JournalforingPanel';
import VelgSak from './VelgSak';
import { Normaltekst } from 'nav-frontend-typografi';
import EkspanderKnapp from '../../../../components/EkspanderKnapp';
import { isLoadedPerson } from '../../../../redux/restReducers/personinformasjon';
import { capitalizeName } from '../../../../utils/stringFormatting';
import AlertStripeInfo from 'nav-frontend-alertstriper/lib/info-alertstripe';
import Select from 'nav-frontend-skjema/lib/select';
import { getSaksbehandlerEnhet } from '../../../../utils/loggInfo/saksbehandlersEnhetInfo';

const FormStyle = styled.form`
    display: flex;
    flex-direction: column;
    align-items: stretch;
    textarea {
        min-height: 9rem;
    }
    button {
        margin-top: 0.5rem;
    }
    .ReactCollapse--collapse .skjemaelement {
        margin: 0;
    }
`;

const tekstMaksLengde = 5000;

enum Oppgaveliste {
    MinListe = 'MinListe',
    EnhetensListe = 'Enhetensliste'
}

function SendNyMelding() {
    const initialDialogType = Meldingstype.SpørsmålSkriftlig;
    const [tema, setTema] = useState<Kodeverk | undefined>(undefined);
    const [sak, setSak] = useState<JournalforingsSak | undefined>(undefined);
    const [tekst, setTekst] = useState('');
    const [dialogType, setDialogType] = useState(initialDialogType);
    const [tekstFeil, setTekstFeil] = useState(false);
    const [temaFeil, setTemaFeil] = useState(false);
    const [oppgaveListe, setOppgaveliste] = useState(Oppgaveliste.MinListe);
    const [visSaker, setVisSaker] = useState(false);
    const [visFeilMeldinger, setVisFeilmeldinger] = useState(false);

    const sammensattesaker = useSelector((state: AppState) => state.restResources.sammensatteSaker);
    const psaksaker = useSelector((state: AppState) => state.restResources.psakSaker);
    const personinformasjon = useSelector((state: AppState) => state.restResources.personinformasjon);
    const dispatch = useDispatch();

    const enhet = getSaksbehandlerEnhet();

    const saker: JournalforingsSak[] =
        isLoaded(sammensattesaker) && isLoaded(psaksaker) ? [...sammensattesaker.data, ...psaksaker.data] : [];

    useEffect(() => {
        setTekstFeil(tekst.length === 0 || tekst.length > tekstMaksLengde);
        setTemaFeil(!tema);
        setVisFeilmeldinger(false);
    }, [tekst, tema]);

    if (isNotStarted(sammensattesaker)) {
        dispatch(sammensattesaker.actions.fetch);
    }

    if (isNotStarted(psaksaker)) {
        dispatch(psaksaker.actions.fetch);
    }

    const erReferat = dialogType !== Meldingstype.SpørsmålSkriftlig;
    const erOppmøte = dialogType !== Meldingstype.SamtalereferatOppmøte;
    const erGyldigReferat = !temaFeil && !tekstFeil;
    const erGyldigSpørsmål = true;

    const handleSubmit = (event: FormEvent) => {
        event.preventDefault();
        if (tema && erReferat && erGyldigReferat) {
            dispatch(
                sendMeldingActionCreator({
                    fritekst: tekst,
                    kanal: erOppmøte ? 'OPPMOTE' : 'TELEFON',
                    type: dialogType,
                    temagruppe: tema.kodeRef,
                    traadId: null,
                    kontorsperretEnhet: null,
                    erTilknyttetAnsatt: true
                })
            );
        } else if (!erReferat && sak && erGyldigSpørsmål) {
            dispatch(
                sendMeldingActionCreator({
                    fritekst: tekst,
                    kanal: '',
                    type: dialogType,
                    temagruppe: sak.temaKode,
                    traadId: null,
                    kontorsperretEnhet: null,
                    erTilknyttetAnsatt: true
                })
            );
        } else {
            setVisFeilmeldinger(true);
        }
    };

    const handleAvbryt = () => {
        setDialogType(initialDialogType);
        setTekst('');
        setTema(undefined);
    };

    const handleVelgSak = (sak: JournalforingsSak) => {
        setSak(sak);
    };

    const tekstFeilmelding: SkjemaelementFeil | undefined =
        visFeilMeldinger && tekstFeil
            ? {
                  feilmelding: tekst.length === 0 ? 'Du må skrive en tekst' : `Maks ${tekstMaksLengde} tegn`
              }
            : undefined;

    const navn = isLoadedPerson(personinformasjon)
        ? capitalizeName(personinformasjon.data.navn.fornavn || '')
        : 'bruker';

    const spørsmålFields = (
        <>
            <div>
                <Normaltekst>Valgt sak: {sak && sak.saksId}</Normaltekst>
                <EkspanderKnapp onClick={() => setVisSaker(!visSaker)} open={visSaker} />
            </div>
            <UnmountClosed isOpened={visSaker}>
                <section>
                    <VelgSak saker={saker} valgtSak={sak} setValgtSak={handleVelgSak} />{' '}
                    <p style={{ display: 'none' }}>
                        /* TODO Denne byttes ut med komponent fra Journalføring når den er ferdig*/
                    </p>
                </section>
            </UnmountClosed>
            <section>
                <Select
                    label="oppgaveliste"
                    value={oppgaveListe}
                    onChange={event => setOppgaveliste(event.target.value as Oppgaveliste)}
                >
                    <option value={Oppgaveliste.MinListe}>Min oppgaveliste</option>
                    <option value={Oppgaveliste.EnhetensListe}>Skal til {enhet} sin oppgaveliste</option>
                </Select>
            </section>
        </>
    );

    return (
        <article>
            <h3 className="sr-only">Send ny melding</h3>
            <FormStyle onSubmit={handleSubmit}>
                <RadioPanelGruppe
                    name="Dialogtype"
                    legend="Velg dialogtype"
                    radios={[
                        { label: 'Samtalereferat telefon', value: Meldingstype.SamtalereferatTelefon },
                        { label: 'Spørsmål til bruker', value: Meldingstype.SpørsmålSkriftlig },
                        { label: 'Samtalereferat oppmøte', value: Meldingstype.SamtalereferatOppmøte }
                    ]}
                    checked={dialogType}
                    onChange={(_, value) => setDialogType(value as Meldingstype)}
                />
                <UnmountClosed isOpened={erReferat}>
                    <Temavelger setTema={setTema} tema={tema} visFeilmelding={temaFeil && visFeilMeldinger} />
                </UnmountClosed>
                <UnmountClosed isOpened={!erReferat}>{spørsmålFields}</UnmountClosed>
                <Textarea
                    value={tekst}
                    onChange={e => setTekst((e as React.KeyboardEvent<HTMLTextAreaElement>).currentTarget.value)}
                    label={'Melding'}
                    maxLength={tekstMaksLengde}
                    feil={tekstFeilmelding}
                    placeholder={`Alt du skriver i denne boksen blir synlig for brukeren når du trykker "Del med ${navn}"`}
                />
                <UnmountClosed isOpened={!erReferat}>
                    <AlertStripeInfo>Bruker kan svare</AlertStripeInfo>
                </UnmountClosed>
                <KnappBase type="hoved" htmlType="submit">
                    Del med {navn}
                </KnappBase>
                <KnappMedBekreftPopup onBekreft={handleAvbryt} popUpTekst="Du vil miste meldingen du har påbegynnt">
                    Avbryt
                </KnappMedBekreftPopup>
            </FormStyle>
        </article>
    );
}

export default SendNyMelding;
