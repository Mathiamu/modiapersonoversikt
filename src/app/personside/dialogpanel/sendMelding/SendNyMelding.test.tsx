import * as React from 'react';
import * as renderer from 'react-test-renderer';
import SendNyMelding from './SendNyMelding';
import TestProvider from '../../../../test/Testprovider';

test('viser send ny melding', () => {
    const dialogPanelBody = renderer.create(
        <TestProvider>
            <SendNyMelding />
        </TestProvider>
    );

    expect(dialogPanelBody.toJSON()).toMatchSnapshot();
});
