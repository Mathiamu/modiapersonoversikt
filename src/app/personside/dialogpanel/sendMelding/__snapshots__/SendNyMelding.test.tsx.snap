// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`viser send ny melding 1`] = `
.c2 label {
  position: absolute !important;
  height: 1px;
  width: 1px;
  overflow: hidden;
  -webkit-clip: rect(1px 1px 1px 1px);
  clip: rect(1px 1px 1px 1px);
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
}

.c3 .ekspanderbartPanel__hode {
  padding: 0.6rem;
}

.c5 textarea {
  min-height: 9rem;
}

.c5 label {
  position: absolute !important;
  height: 1px;
  width: 1px;
  overflow: hidden;
  -webkit-clip: rect(1px 1px 1px 1px);
  clip: rect(1px 1px 1px 1px);
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
}

.c1 label {
  position: absolute !important;
  height: 1px;
  width: 1px;
  overflow: hidden;
  -webkit-clip: rect(1px 1px 1px 1px);
  clip: rect(1px 1px 1px 1px);
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
}

.c4 {
  margin-top: 1rem;
}

.c4 label {
  position: absolute !important;
  height: 1px;
  width: 1px;
  overflow: hidden;
  -webkit-clip: rect(1px 1px 1px 1px);
  clip: rect(1px 1px 1px 1px);
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-top: 1rem;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
}

.c0 .skjemaelement {
  margin-bottom: 0;
}

.c0 > * {
  margin-bottom: 1rem;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c6 > * {
  margin-bottom: 0.7rem;
  margin-right: 0.5rem;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

<article>
  <h2
    className="typo-undertittel"
  >
    Send ny melding
  </h2>
  <form
    className="c0"
    onSubmit={[Function]}
  >
    <div>
      <div
        className="skjemaelement c1"
      >
        <label
          className="skjemaelement__label"
          htmlFor="Helt tilfeldig ID"
        >
          Velg dialogtype
        </label>
        <div
          className="selectContainer input--fullbredde"
        >
          <select
            className="skjemaelement__input"
            disabled={false}
            id="Helt tilfeldig ID"
            onChange={[Function]}
            value="SAMTALEREFERAT_TELEFON"
          >
            <option
              value="SAMTALEREFERAT_TELEFON"
            >
              Samtalereferat telefon
            </option>
            <option
              value="SAMTALEREFERAT_OPPMOTE"
            >
              Samtalereferat oppmøte
            </option>
            <option
              value="SPORSMAL_MODIA_UTGAAENDE"
            >
              Spørsmål til bruker
            </option>
          </select>
        </div>
        <div
          aria-live="assertive"
          role="alert"
        />
      </div>
    </div>
    <div
      className=""
    >
      <div
        className=""
      >
        <div
          className="skjemaelement c2"
        >
          <label
            className="skjemaelement__label"
            htmlFor="Helt tilfeldig ID"
          >
            Tema
          </label>
          <div
            className="selectContainer input--fullbredde"
          >
            <select
              className="skjemaelement__input"
              disabled={false}
              id="Helt tilfeldig ID"
              onChange={[Function]}
              value=""
            >
              <option
                disabled={true}
                value=""
              >
                Velg temagruppe
              </option>
              <option
                value="ARBD"
              >
                Arbeid
              </option>
              <option
                value="FMLI"
              >
                Familie
              </option>
              <option
                value="HJLPM"
              >
                Hjelpemiddel
              </option>
              <option
                value="PENS"
              >
                Pensjon
              </option>
              <option
                value="OVRG"
              >
                Øvrig
              </option>
            </select>
          </div>
          <div
            aria-live="assertive"
            role="alert"
          />
        </div>
        <div
          aria-live="assertive"
          role="alert"
        />
      </div>
      <div
        className=""
      >
        <div
          className="ekspanderbartPanel c3 ekspanderbartPanel--lukket ekspanderbartPanel--border"
        >
          <button
            aria-expanded={false}
            className="ekspanderbartPanel__hode"
            onClick={[Function]}
            onKeyDown={[Function]}
            type="button"
          >
            <div
              className="ekspanderbartPanel__flex-wrapper"
            >
              <span
                className="typo-normal ekspanderbartPanel__heading"
              >
                Velg sak
              </span>
              <span
                className="ekspanderbartPanel__indikator"
              />
            </div>
          </button>
          <article
            aria-label="Velg sak"
            className="ekspanderbartPanel__innhold"
          >
            <svg
              aria-label="Laster innhold"
              className="spinner spinner--xl"
              focusable="false"
              height={32}
              kind="spinner"
              preserveAspectRatio="xMidYMid"
              viewBox="0 0 50 50"
              width={32}
            >
              <title>
                Venter...
              </title>
              <circle
                cx="25"
                cy="25"
                fill="none"
                r="20"
                stroke="#bdbab7"
                strokeWidth="5"
                xmlns="http://www.w3.org/2000/svg"
              />
              <circle
                cx="25"
                cy="25"
                fill="none"
                r="20"
                stroke="#7f756c"
                strokeDasharray="50 155"
                strokeLinecap="round"
                strokeWidth="5"
              />
            </svg>
          </article>
        </div>
        <div
          aria-live="assertive"
          role="alert"
        />
      </div>
      <div
        className="skjemaelement c4"
      >
        <label
          className="skjemaelement__label"
          htmlFor="Helt tilfeldig ID"
        >
          Oppgaveliste
        </label>
        <div
          className="selectContainer input--fullbredde"
        >
          <select
            className="skjemaelement__input"
            disabled={false}
            id="Helt tilfeldig ID"
            onChange={[Function]}
            value="MinListe"
          >
            <option
              value="MinListe"
            >
              Svar skal til min oppgaveliste
            </option>
            <option
              value="Enhetensliste"
            >
              Svar skal til 
               sin oppgaveliste
            </option>
          </select>
        </div>
        <div
          aria-live="assertive"
          role="alert"
        />
      </div>
    </div>
    <div
      className="c5"
    >
      <div
        className="skjemaelement textarea__container"
      >
        <label
          className="skjemaelement__label"
          htmlFor="Helt tilfeldig ID"
        >
          Melding
        </label>
        <div
          className="textarea--medMeta__wrapper"
        >
          <textarea
            className="skjemaelement__input textarea--medMeta"
            id="Helt tilfeldig ID"
            onChange={[Function]}
            placeholder="Alt du skriver i denne boksen blir synlig for Aremark når du trykker \\"Del med Aremark\\""
            style={
              Object {
                "height": "30px",
              }
            }
            type="text"
            value=""
          />
          <p
            className="textarea--medMeta__teller"
          >
            <span>
              Du har 
              5000
               tegn igjen
            </span>
          </p>
        </div>
        <div
          aria-live="assertive"
          role="alert"
        />
        <div
          aria-hidden="true"
          className="textareamirror"
        />
      </div>
      <div
        aria-live="assertive"
        role="alert"
      />
    </div>
    <div
      className=""
    >
      <div
        className="alertstripe alertstripe--info"
      >
        <span
          aria-label="info"
          className="alertstripe__ikon"
        >
          <svg
            focusable="false"
            height="1.5em"
            kind="info-sirkel-fyll"
            viewBox="0 0 24 24"
            width="1.5em"
          >
            <g
              fill="none"
              fillRule="nonzero"
            >
              <path
                d="M12 0C5.382 0 0 5.382 0 12s5.382 12 12 12c6.617 0 12-5.382 12-12S18.617 0 12 0z"
                fill="#669DB4"
              />
              <g
                fill="#FFF"
              >
                <path
                  d="M12 5a1.566 1.566 0 1 1 .11 3.13A1.566 1.566 0 0 1 12 5zM14.976 17.01c.563 0 1.043.431 1.043.991s-.48.992-1.043.992H9.39c-.564 0-1.043-.431-1.043-.992 0-.56.479-.99 1.043-.99h1.6v-5.016h-.986c-.565 0-1.044-.43-1.044-.991 0-.56.48-.991 1.044-.991h2.03c.563 0 1.043.43 1.043.99v6.007H14.976z"
                />
              </g>
            </g>
          </svg>
        </span>
        <div
          className="typo-normal alertstripe__tekst"
        >
          Bruker kan svare
        </div>
      </div>
    </div>
    <div
      className="c6"
    >
      <button
        className="knapp knapp--hoved"
        disabled={false}
        type="submit"
      >
        Del med 
        Aremark
      </button>
      <button
        className="knapp knapp--flat"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Avbryt
      </button>
    </div>
  </form>
</article>
`;
